precompiles/assets-erc20/src/lib.rs

```diff
@@ -0,0 +1,444 @@
+use frame_support::traits::fungibles::Inspect;
+use frame_support::traits::OriginTrait;
+use frame_support::{
+	dispatch::{Dispatchable, GetDispatchInfo, PostDispatchInfo},
+	sp_runtime::traits::StaticLookup,
+};
+
+use pallet_evm::{AddressMapping, Precompile};
+use precompile_utils::{
+	error, Address, EvmDataReader, EvmDataWriter, EvmResult, Gasometer, LogsBuilder, RuntimeHelper,
+};
+
+use slices::u8_slice;
+use sp_core::{H160, U256};
+use sp_std::{
+	convert::{TryFrom, TryInto},
+	marker::PhantomData,
+	vec,
+};
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+/// Solidity selector of the Transfer log, which is the Keccak of the Log signature.
+pub const SELECTOR_LOG_TRANSFER: &[u8; 32] =
+	u8_slice!("0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef");
+
+/// Solidity selector of the Approval log, which is the Keccak of the Log signature.
+pub const SELECTOR_LOG_APPROVAL: &[u8; 32] =
+	u8_slice!("0x8c5be1e5ebec7d5bd14f71427d1e84f3dd0314c0f7b2291e5b200ac8c7c3b925");
+
+/// Alias for the Balance type for the provided Runtime and Instance.
+pub type BalanceOf<Runtime, Instance = ()> = <Runtime as pallet_assets::Config<Instance>>::Balance;
+
+/// Alias for the Asset Id type for the provided Runtime and Instance.
+pub type AssetIdOf<Runtime, Instance = ()> = <Runtime as pallet_assets::Config<Instance>>::AssetId;
+
+#[precompile_utils::generate_function_selector]
+#[derive(Debug, PartialEq, num_enum::TryFromPrimitive, num_enum::IntoPrimitive)]
+pub enum Action {
+	TotalSupply = "totalSupply()",
+	BalanceOf = "balanceOf(address)",
+	Allowance = "allowance(address,address)",
+	Transfer = "transfer(address,uint256)",
+	Approve = "approve(address,uint256)",
+	TransferFrom = "transferFrom(address,address,uint256)",
+}
+
+/// This trait ensure we can convert AccountIds to AssetIds
+/// We will require Runtime::Precompiles to have this trait implemented
+pub trait AccountIdToAssetId<Account, AssetId> {
+	// Get assetId from account
+	fn account_to_asset_id(account: Account) -> Option<AssetId>;
+}
+
+/// Precompile exposing a pallet_assets as an ERC20.
+/// Multiple precompiles can support instances of pallet_assetss.
+/// The precompile uses an additional storage to store approvals.
+pub struct Erc20AssetsPrecompile<Runtime, Instance: 'static = ()>(PhantomData<(Runtime, Instance)>);
+
+impl<Runtime, Instance> Precompile for Erc20AssetsPrecompile<Runtime, Instance>
+where
+	Instance: 'static,
+	Runtime: pallet_assets::Config<Instance> + pallet_evm::Config,
+	Runtime::Call: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+	Runtime::Call: From<pallet_assets::Call<Runtime, Instance>>,
+	<Runtime::Call as Dispatchable>::Origin: From<Option<Runtime::AccountId>>,
+	BalanceOf<Runtime, Instance>: TryFrom<U256> + Into<U256>,
+	Runtime::Precompiles: AccountIdToAssetId<Runtime::AccountId, AssetIdOf<Runtime, Instance>>,
+	<<Runtime as frame_system::Config>::Call as Dispatchable>::Origin: OriginTrait,
```

Comment:
This is a trait bound on `Runtime as pallet_evm::Config>::Precompiles` right? So the precompiles type has to implement this trait?

Reply:
This seems like the wrong type to require this trait bound on. For one thing, this trait is only useful for this particular precopmile. Another thing: doing it this way means that You can't ship a "reference implementation" with the precompile because it always has to be implemented on a particular runtime's `PrecompileSet`.

I'd recommend either:
1. Don't make this conversion customizeable, just hard-code it.
2. Add a type parameter to `Erc20AssetsPrecompile` with this trait bound. (and phantom data).

Reply:
Having now read the entire PR, I recommend that you call your trait, `IntoAssetId` and use the trait bound `Runtime::AccountId: Into<AssetId>`.

Then you impl `IntoAssetId` for `H160`.

Reply:
Perfect

Reply:
Your last idea wont work since I need at least the trait (which should leave in erc20asset precompiles, which is the one requiring the bound), or one of the types to be declared in the module in which I am implementing the trait, which I would like it to be the moonbase runtime. So I think I am going to opt for creating a new type parameter to  Erc20AssetsPrecompile that is able to convert between both.

Reply:
as per @nanocryk's suggestion, I changed the asset precompile to implement precompileSet in addition to Precompile. This matches better the current design, as asset holds a set of precompiles and not just one precompile. Inside this trait implementation, I am running the necessary checks related to the AssetId through the AccountIdToAssetId bound, which is now a bound required to the runtime and not to MoonbasePrecompiles. I think its much cleaner now, let me know what you think!

Reply:
Yeah, that's a great idea.